src/main/java/com/google/devtools/build/lib/vfs/DigestHashFunction.java

```diff
@@ -67,6 +68,7 @@ public int getDigestMaximumLength() {
 
   public static final DigestHashFunction SHA1 = register(Hashing.sha1(), "SHA-1", "SHA1");
   public static final DigestHashFunction SHA256 = register(Hashing.sha256(), "SHA-256", "SHA256");
+  public static final DigestHashFunction BLAKE3 = register(new Blake3HashFunction(), "BLAKE3");
```

Comment:
I think you can just use `new MessageDigestHashFunction("BLAKE3")` ([doc](https://cs.opensource.google/guava/guava/+/master:guava/src/com/google/common/hash/MessageDigestHashFunction.java;l=38;drc=e62d6a0456420d295089a9c319b7593a3eae4a83)) instead of implementing `HashFunction` by yourself in `Blake3HashFunction`.

Reply:
As mentioned above, we don't want put BLAKE3 code in the common code path. Neither the registration here nor the provider below.

I suggest we have a utility class `BazelHashFunctions` which installs the provider and register the hash function, e.g.:

```
class BazelHashFunctions {
    static {
        addProvider(...);
    }

    public static final DigestHashFunction BLAKE3 = DigestHashFunction.register(...);
}
```

In a following PR when we want to create a file system based on BLAKE3, we could do that in `BazelFileSystemModule` with code like:

```
new UnixFileSystem(BazelHashFunctions.BLAKE3, ...);
```

Reply:
I'm trying to read between the lines here and figure out where you want stuff / how you want it to be injected -- let me know what I got wrong.

I moved blake3 stuff into a `bazel` subdir of `vfs`. The provider is statically registered when BazelHashFunctions is loaded by the classloader in `src/main/java/com/google/devtools/build/lib/bazel/Bazel.java`. Re the MessageDigestHashFunction -- addressed above, but that's not public afaict.

Reply:
Yes, we want the provider is statically registered by the classloader when it is loading class `BazelHashFunctions`. But we don't want to have it implementing `BlazeModule` and put it in `src/main/java/com/google/devtools/build/lib/bazel/Bazel.java`.

What I proposed above is we could add the code 

```
new UnixFileSystem(BazelHashFunctions.BLAKE3, ...);
```

to somewhere like [here](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/runtime/BazelFileSystemModule.java;l=51;drc=cde925dba0202a7c66458d9e838708ee535994ff) so that when we want to use BLAKE3, it registers it.

We don't need to do it in this CL as we haven't focused on integrating BLAKE3 into Bazel's VFS. Let's limit the scope of this CL to only adapt BLAKE3 as a DigestHashFunction.

Reply:
Can you give a little more context on what else is required to add the digest function to vfs? I'm a little unclear on that part and it will help me for the next cl. Thanks!

Reply:
Well, I think I overcomplicated that. The only necessary part is we want to register `BLAKE3` for Bazel only. We can just force loading `BazelHashFunctions`  by adding following code block in `BazelFilesystemModule`

```
class BazelFilesystemmodule {
    static {
        BazelHashFunctions.ensureRegistered();
    }
}
```

where `BazelHashFunctions.ensureRegistered` is just an empty static function whose sole purpose is to let JVM load class `BazelHashFunctions` and run the static block there.

Reply:
Thanks, this is just what I needed. Done.